platform/mellanox/install-pending-fw.py

```diff
@@ -0,0 +1,96 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+
+from collections import namedtuple
+
+from fwutil.lib import ComponentStatusProvider, PlatformComponentsParser
+from sonic_platform.component import ComponentCPLD, MPFAManager
+
+# Globals
+FW_STATUS_SCHEDULED = "scheduled"
+CPLD_FLAG = False
+
+# Init platform chassis helper classes
+csp = ComponentStatusProvider()
+pcp = PlatformComponentsParser(csp.is_modular_chassis())
+
+# Parse update status file
+update_status = csp.read_au_status_file_if_exists()
+
+if update_status is None:
+    exit(0)
+
+# Parse platform components file
+try:
+    pcp.parse_platform_components()
+except Exception as e:
+    print("Error parsing platform components. Firmware update failed: {}".format(str(e)))
+    print("System will reboot in 10 seconds please fix issue and run update command again.")
+    time.sleep(10)
+    exit(-1)
+
+# Iterate each component in the status file
+comp_install = []
+files = []
+
+for boot_type, components in update_status.items():
+    for comp in components:
+
+        # Skip if fw isn't scheduled for install at reboot
+        if comp["info"] != FW_STATUS_SCHEDULED: continue
+
+        # Get component object and target firmware file
+        key = comp["comp"]
+        comp_path = key.split("/")
+
+        if len(comp_path) == 3:
+            # Module component
+            _, parent_name, comp_name = comp_path
+            fw_file = pcp.module_component_map[parent_name][comp_name]["firmware"]
+            component = csp.module_component_map[parent_name][comp_name]
+        else: 
+            # Chassis component
+            parent_name, comp_name = comp_path
+            fw_file = pcp.chassis_component_map[parent_name][comp_name]["firmware"]
+            component = csp.chassis_component_map[parent_name][comp_name]
+
+        # Install firmware. If CPLD flag to be installed last due to force reboot during refresh
+        if type(component) == ComponentCPLD:
+            if CPLD_FLAG:
+                print("WARNING: Multiple CPLD firmwares defined. Some CPLD updates may not fully complete.")
+            CPLD_FLAG = True
+            mpfa = MPFAManager(fw_file)
+            mpfa.extract()
+            if not mpfa.get_metadata().has_option('firmware', 'refresh'):
+                raise RuntimeError("Failed to get {} refresh firmware".format(self.name))
+            refresh_firmware = mpfa.get_metadata().get('firmware', 'refresh')
+            comp_install = comp_install + [component]
+            files = files + [os.path.join(mpfa.get_path(), refresh_firmware)]
+        else:
+            comp_install = [component] + comp_install
+            files = [fw_file] + files
+
+# Do install
+for i, c in enumerate(comp_install):
+    if type(c) == ComponentCPLD:
+        c.install_firmware(files[i])
+    else:
+        c.install_firmware(files[i], allow_reboot=False)
```

Comment:
@alexrallen what's the idea behind of two lists here? Why not using a list of tuples?

Reply:
Personal preference, I like to avoid tuples unless they are named tuples to make code more readable. I was on the fence here but I chose the two lists over a single list of named tuples because I thought it was a bit simpler.

Reply:
@alexrallen understood